pkg/standalone-utils/test/PermitRelayer.test.ts

```diff
@@ -0,0 +1,432 @@
+
+import Token from '@balancer-labs/v2-helpers/src/models/tokens/Token';
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import { GeneralPool } from '@balancer-labs/v2-helpers/src/models/vault/pools';
+import { encodeExit, encodeJoin } from '@balancer-labs/v2-helpers/src/models/pools/mockPool';
+import {
+  encodeCalldataAuthorization,
+  signBatchSwapAuthorization,
+  signExitAuthorization,
+  signJoinAuthorization,
+  signPermit,
+  signSwapAuthorization,
+} from '@balancer-labs/v2-helpers/src/models/misc/signatures';
+
+import { deploy } from '@balancer-labs/v2-helpers/src/contract';
+import * as expectEvent from '@balancer-labs/v2-helpers/src/test/expectEvent';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+
+import { bn } from '@balancer-labs/v2-helpers/src/numbers';
+import { MAX_UINT256, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+
+const tokenInitialBalance = bn(200e18);
+
+const setup = async () => {
+  const [, admin, lp] = await ethers.getSigners();
+
+  const dai = await deploy('TestTokenPermitDAI', { args: [admin.address, 'DAI', 'DAI', 18] });
+  const mkr = await deploy('TestTokenPermit', { args: [admin.address, 'MKR', 'MKR', 18] });
+
+  const tokenList = [await Token.fromContract(dai), await Token.fromContract(mkr), await Token.create('WETH')];
+  const tokens = new TokenList(tokenList.sort((a, b) => a.compare(b)));
+
+  // Deploy Balancer Vault
+  const authorizer = await deploy('v2-vault/Authorizer', { args: [admin.address] });
+  const vault = await deploy('v2-vault/Vault', { args: [authorizer.address, tokens.WETH.address, 0, 0] });
+
+  // Deploy Pool
+  const pool = await deploy('v2-vault/test/MockPool', { args: [vault.address, GeneralPool] });
+  const poolId = await pool.getPoolId();
+
+  await tokens.mint({ to: lp, amount: tokenInitialBalance.mul(2), from: admin });
+  await tokens.approve({ to: vault.address, from: [lp] });
+
+  await pool.registerTokens(tokens.addresses, [ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS]);
+
+  await vault.connect(lp).joinPool(poolId, lp.address, lp.address, {
+    assets: tokens.addresses,
+    maxAmountsIn: tokens.addresses.map(() => MAX_UINT256),
+    fromInternalBalance: false,
+    userData: encodeJoin(
+      tokens.addresses.map(() => tokenInitialBalance),
+      tokens.addresses.map(() => 0)
+    ),
+  });
+
+  await tokens.approve({ to: vault.address, from: [lp], amount: 0 });
+
+  const relayer = await deploy('PermitRelayer', { args: [vault.address] });
+
+  // Authorise relayer to perform actions.
+  const single = await actionId(vault, 'swap');
+  const batch = await actionId(vault, 'batchSwap');
+  const manageUserBalance = await actionId(vault, 'manageUserBalance');
+  const joinPool = await actionId(vault, 'joinPool');
+  const exitPool = await actionId(vault, 'exitPool');
+  await authorizer.connect(admin).grantRoles([single, batch, manageUserBalance, joinPool, exitPool], relayer.address);
```

Comment:
I'd try to split up this very large setup into smaller parts, as we'd likely want to have at least some tests for partial setup (e.g. what happens if the contract has not been authorized), using `context` and `sharedBeforeEach`